parser/parse.go

```diff
@@ -239,77 +245,100 @@ func (b *Builder) AddDirRecursive(dir string) error {
 // GOPATH, GOROOT, and the location of your go binary (`which go`) will all be
 // searched if dir doesn't literally resolve.
 func (b *Builder) AddDirTo(dir string, u *types.Universe) error {
-	if _, found := b.parsed[dir]; !found {
-		// We want all types from this package, as if they were directly added
-		// by the user.  They WERE added by the user, in effect.
-		if err := b.addDir(dir, true); err != nil {
-			return err
-		}
-	} else {
-		// We already had this package, but we want it to be considered as if
-		// the user addid it directly.
-		b.userRequested[dir] = true
+	// We want all types from this package, as if they were directly added
+	// by the user.  They WERE added by the user, in effect.
+	if _, err := b.importPackage(dir, true); err != nil {
+		return err
 	}
-	return b.findTypesIn(dir, u)
+	return b.findTypesIn(importPathString(b.buildPackages[dir].ImportPath), u)
 }
 
 // The implementation of AddDir. A flag indicates whether this directory was
 // user-requested or just from following the import graph.
 func (b *Builder) addDir(dir string, userRequested bool) error {
-	pkg, err := b.buildPackage(dir)
+	glog.V(5).Infof("addDir %s", dir)
+	buildPkg, err := b.importBuildPackage(dir)
 	if err != nil {
 		return err
 	}
-	// Check in case this package was added (maybe dir was not canonical)
-	if wasRequested, wasAdded := b.userRequested[dir]; wasAdded {
-		if !userRequested || userRequested == wasRequested {
-			return nil
+	pkgPath := importPathString(buildPkg.ImportPath)
```

Comment:
Yeah, this it here.  The `ImportPath` is *sometimes*, but not always the same as package name.  It seems to be intentionally not a valid package path based on construction here https://golang.org/src/go/build/build.go#L570 .  Still thinking about how to detect it.